### PR TITLE
Oop/nh pgrad

### PR DESCRIPTION
**Pull request by oelbert**
## Purpose

This PR changes the implementation of `nh_p_grad` into an object-oriented one.

## Code changes:

- `nh_p_grad` is now accessed through a `NonHydrostaticPressureGradient` class. Calls in the test class and `dyn_core` have been updated. 

## Checklist
Before submitting this PR, please make sure:

- [x] You have followed the coding standards guidelines established at [Code Review Checklist](https://paper.dropbox.com/doc/Code-Review-Checklist--BD7zigBMAhMZAPkeNENeuU2UAg-IlsYffZgTwyKEylty7NhY).
- [x] Docstrings and type hints are added to new and updated routines, as appropriate
- [x] All relevant documentation has been updated or added (e.g. README, CONTRIBUTING docs)
- [x] Unit tests are added or updated for non-stencil code changes

### CODE REVIEW

File: fv3core/stencils/nh_p_grad.py [hunk 1]
```diff
@@ -1,26 +1,23 @@
...
 import fv3core._config as spec
 import fv3core.stencils.a2b_ord4 as a2b_ord4
+import fv3core.utils.global_config as global_config
 import fv3core.utils.gt4py_utils as utils
-from fv3core.decorators import gtstencil
 from fv3core.utils.typing import FloatField, FloatFieldIJ
 
 
-@gtstencil()
 def set_k0(pp: FloatField, pk3: FloatField, top_value: float):
```
Comment (twicki) [inline fv3core/stencils/nh_p_grad.py:10]: if we are already in the process, can we make the name of this consistent with the other stencils?

File: fv3core/stencils/nh_p_grad.py [hunk 2]
```diff
@@ -90,60 +86,129 @@ def CalcV(
...
+        u=u v=v pp=pkc gz=gz pk3=pk3 delp=delp dt=dt
+        """
+        ptk = ptop ** akap
+        top_value = ptk  # = peln1 if spec.namelist.use_logp else ptk
+
+        self._set_k0_stencil(
+            pp, pk3, top_value, origin=self.orig, domain=self.domain_k1
+        )
+
+        a2b_ord4.compute(pp, self._tmp_wk1, kstart=1, nk=self.nk, replace=True)
```
Comment (twicki) [inline fv3core/stencils/nh_p_grad.py:177]: what is the strategy here? go flat and only change the immediate calls?

File: fv3core/stencils/nh_p_grad.py [hunk 3]
```diff
@@ -90,60 +86,129 @@ def CalcV(
...
+        pk3: FloatField,
+        delp: FloatField,
+        dt: float,
+        ptop: float,
+        akap: float,
+    ):
+        """
+        Updates the U and V winds due to pressure gradients,
+        accounting for both the hydrostatic and nonhydrostatic contributions.
+        Args:
```
Comment (twicki) [inline fv3core/stencils/nh_p_grad.py:157]: should we split between input and inout args to get a sense of data flow?

File: fv3core/stencils/nh_p_grad.py [hunk 4]
```diff
@@ -90,60 +86,129 @@ def CalcV(
...
+        )
+
+        self._calc_wk_stencil = stencil(
+            definition=CalcWk,
+            backend=global_config.get_backend(),
+            rebuild=global_config.get_rebuild(),
+        )
+
+        self._calc_u_stencil = stencil(
+            definition=CalcU,
```
Comment (rheacangeo) [inline fv3core/stencils/nh_p_grad.py:131]: most stencils don't use campy camel style for their definition -- we could take refactor opportunity to have these match, but if you prefer not to that is fine

File: fv3core/stencils/nh_p_grad.py (hunk 2 above)
Comment (oelbert) [reply fv3core/stencils/nh_p_grad.py:177]: Correct. The a2b_ord4 calls should be changed when it's refactored

File: fv3core/stencils/nh_p_grad.py (hunk 1 above)
Comment (rheacangeo) [reply fv3core/stencils/nh_p_grad.py:10]: I think we are only using campy camel for class names, I don't think there's many if any other places in the code we use it for stencil definitions -- we could though! Maybe I missed the memo though?

File: fv3core/stencils/nh_p_grad.py (hunk 3 above)
Comment (oelbert) [reply fv3core/stencils/nh_p_grad.py:157]: Could do. Even though we return u, v, pp, gz, pk3, and delp I think only u and v are updated, so we could maybe stop returning the other variables

File: fv3core/stencils/nh_p_grad.py [hunk 5]
```diff
@@ -90,60 +86,129 @@ def CalcV(
         ) * rdy
 
 
-def compute(u, v, pp, gz, pk3, delp, dt, ptop, akap):
+class NonHydrostaticPressureGradient:
     """
-    u=u v=v pp=pkc gz=gz pk3=pk3 delp=delp dt=dt
+    Fortran name is nh_p_grad
```
[the remark this answers is not in the document]
Comment (oelbert) [reply fv3core/stencils/nh_p_grad.py:91]: I agree with not adding test-specific code to the model code, and grepping for the class name or filename seems ok to me?

File: fv3core/stencils/nh_p_grad.py (hunk 4 above)
Comment (oelbert) [reply fv3core/stencils/nh_p_grad.py:131]: good point!

File: fv3core/stencils/nh_p_grad.py [hunk 6]
```diff
@@ -90,60 +86,129 @@ def CalcV(
...
-    )
-    return u, v, pp, gz, pk3, delp
+
+    def __init__(self):
+        grid = spec.grid
+        self.orig = (grid.is_, grid.js, 0)
+        self.domain_full_k = (grid.nic + 1, grid.njc + 1, grid.npz)
+        self.domain_k1 = (grid.nic + 1, grid.njc + 1, 1)
+        self.u_domain = (grid.nic, grid.njc + 1, grid.npz)
+        self.v_domain = (grid.nic + 1, grid.njc, grid.npz)
```
[the remark this answers is not in the document]
Comment (rheacangeo) [reply fv3core/stencils/nh_p_grad.py:100]: ** except the one with k=1 :)

**Pull request merged**